src/solvers/dg/3d/dg.jl

```diff
@@ -1307,11 +1319,120 @@ function rhs!(dg::Dg3D, t_stage)
+# Calculate volume integral and update u_t
+apply_positivity_preserving_limiter!(dg::AbstractDg) = apply_positivity_preserving_limiter!(dg.elements.u, dg, equations(dg))
+
+# Only implemented and tested for compressible Euler equations in 3D
+function apply_positivity_preserving_limiter!(u, dg::AbstractDg, equation::AbstractEquation)
+end
+
+# Apply positivity limiter of Zhan and Shu to nodal values elements.u
+function apply_positivity_preserving_limiter!(u, dg::Dg3D, equation::CompressibleEulerEquations3D)
+  if dg.pp_limiter_apply
+    apply_positivity_preserving_limiter_density!(u, dg::Dg3D, equation::CompressibleEulerEquations3D)
+    apply_positivity_preserving_limiter_pressure!(u, dg::Dg3D, equation::CompressibleEulerEquations3D)
+  end
+end
+
+# Apply positivity limiter of Zhan and Shu to nodal values elements.u
+# density
+function apply_positivity_preserving_limiter_density!(u, dg::Dg3D, equation::CompressibleEulerEquations3D)
+  @unpack weights, pp_limiter_threshold = dg
+  Threads.@threads for element_id in 1:dg.n_elements
+    # Dermine minimum density
+    rho_min = 0.0
+    for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+      u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+      rho_min = min(rho_min, u_node[1])
+    end
+    U1_mean = 0.0
+    U2_mean = 0.0
+    U3_mean = 0.0
+    U4_mean = 0.0
+    U5_mean = 0.0
+    for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+      u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+      U1_mean += u_node[1] * weights[i] * weights[j] * weights[k] / 8.0
+      U2_mean += u_node[2] * weights[i] * weights[j] * weights[k] / 8.0 
+      U3_mean += u_node[3] * weights[i] * weights[j] * weights[k] / 8.0
+      U4_mean += u_node[4] * weights[i] * weights[j] * weights[k] / 8.0
+      U5_mean += u_node[5] * weights[i] * weights[j] * weights[k] / 8.0
+    end
+    # Detect if limiting is necessary
+    Theta_rho = 1
+    if rho_min < pp_limiter_threshold 
+      Theta_rho = (U1_mean - pp_limiter_threshold) / (U1_mean - rho_min)
+    end
+    Theta = min(1, Theta_rho)
+    # Apply limiter if necessary
+    if Theta < 1
+      #println("apply PP limiter 1 ", Theta_rho)
+      for k in 1:nnodes(dg), j in 1:nnodes(dg), i in 1:nnodes(dg)
+        u_node = get_node_vars(dg.elements.u, dg, i, j, k, element_id)
+        dg.elements.u[1,i,j,k,element_id] = Theta * u_node[1] + (1-Theta) * U1_mean 
+        dg.elements.u[2,i,j,k,element_id] = Theta * u_node[2] + (1-Theta) * U2_mean 
+        dg.elements.u[3,i,j,k,element_id] = Theta * u_node[3] + (1-Theta) * U3_mean 
+        dg.elements.u[4,i,j,k,element_id] = Theta * u_node[4] + (1-Theta) * U4_mean 
+        dg.elements.u[5,i,j,k,element_id] = Theta * u_node[5] + (1-Theta) * U5_mean 
+      end
+    end
+  end
+end
+
+# Apply positivity limiter of Zhan and Shu to nodal values elements.u
+# pressure
+function apply_positivity_preserving_limiter_pressure!(u, dg::Dg3D, equation::CompressibleEulerEquations3D)
```

Comment:
`apply_positivity_preserving_limiter_density!` and `apply_positivity_preserving_limiter_pressure!` seem to be more or less the same, except for the choice of which scalar-valued function should be applied to the conservative variables (density or pressure). What about writing a single implementation of the limiter which gets an additional argument such as
```julia
function apply_positivity_preserving_limiter!(func, u, dg::Dg3D, equation)
```
where `func` is called as `scalar = func(u_node, equation)`? Then, we could implement the functions
```julia
function density(u, equation::CompressibleEulerEquations3D)
  rho, rho_v1, rho_v2, rho_v3, rho_e = u
  return rho
end

function pressure(u, equation::CompressibleEulerEquations3D)
  rho, rho_v1, rho_v2, rho_v3, rho_e = u
  return (equation.gamma-1) * (rho_e - 0.5 * (u_noderho_v1^2 + rho_v2^2 + rho_v3^2) / rho)
end
```
These function can also be useful for #143.

Reply:
I like this idea

Reply:
Me too.

Reply:
Also, all the changes and suggestions for the `_density` version also apply to the `_pressure` version.

Reply:
Done.